week-5-e

rowell-17 opened the pull request:
Initial Notes. (NOT FULLY FINISHED). Everything up until the main is carefully written.

rowell-17 committed:
Create week-5-e-SCvx-and-GuSTO-SCP.md

Initial commit

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

Added Motivation and Introduction

rowell-17 committed:
Add files via upload

SCP Block Diagram

rowell-17 committed:
Add files via upload

SCP Hierarchy

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

Worked on Preliminaries

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

Initial Pull Request

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+they are used for and how to use them. This is a collection of notes taken from a paper written by researchers from the University of 
+Washington's Autonomous Controls Laboratory (UW ACL), the developers of the two algorithms[1].
+
+## Motivation and Introduction
```
The motivation and Introduction section has a clear structure and effectively uses diagrams, making them visually easy to understand.

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+![SCP Block Diagram](figs/SCP_diagram.png)
+**Figure 1:** Block diagram of a typical SCP algorithm acquired from [1].
+
+An algorithm hierarchy in **Fig. 2** shows where SCP lies. Typically, the convex solver in layer 2 of **Fig. 1** iteratively calls an **Interior Point Method (IPM)** algorithm from layer 3 in **Fig. 2** that solves a convex problem with linear equality and convex inequality constraints as a sequence of linear equality-constrained problems and thus, IPM iteratively calls the algorithm in layer 2 i ...
```
What is the layer 2 of Fig. 1? Is it the layer 2 of Fig. 2?

rowell-17 replied:
The layers are denoted by the circled numbers. Layer 2 of SCP calls an IPM algorithm, the layer 3 of Fig. 2. Then the IPM algorithm calls layer 2 of Fig. 2.

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+## Preliminaries
+**Formulation**
+The objective of SCP is to solve continuous-time-optimal control problems in the following form:
+|Equation  |#      |
+|----------|-------|
+|$$\min_{u, p} J(x, u, p),$$| (1a)|
+|$$s.t. \quad \dot{x}(t) = f(t, x(t), u(t), p),$$| (1b)|
+|$$(x(t), p) \in \mathcal{X}(t),$$| (1c)|
+|$$(u(t), p) \in \mathcal{U}(t),$$| (1d)|
+|$$s(t, x(t), u(t), p) \leq 0,$$| (1e)|
+|$$g_{ic}(x(0), p) = 0,$$| (1f)|
+|$$g_{tc}(x(1), p) = 0,$$| (1g)|
```
The space between equations is a bit too wide, making them somewhat difficult to read. Please consider using the following markdown notation instead:
```
$$
\begin{align}
\min_{u, p}& J(x, u, p) \\ 
s.t. \quad &\dot{x}(t) = f(t, x(t), u(t), p) \\
&(x(t), p) \in \mathcal{X}(t)\\
&(u(t), p) \in \mathcal{U}(t)\\
&s(t, x(t), u(t), p) \leq 0\\
&g_{ic}(x(0), p) = 0\\
&g_{tc}(x(1), p) = 0
\end{align}
$$
```

rowell-17 replied:
equation format accepted. I added labels using \tag{label}

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+|$$H_{0}x(0) + K_{0}p + l_{0},$$| (6f)|
+|$$H_{f}x(1) + K_{f}p + l_{f} = 0.$$| (6g)|
+
+Notice that the cost is not convexified while everything else is convex in (6). The cost function will be convexified through SCvx and GuSTO algorithm to be discussed. The control trajectory $u(.)$ is an infinite-dimensional vector space of continuous-time functions and needs to be discretized through temporal discretization and direct colocation so it can be solved numerically.
```
It might be better to write $u(\cdot)$ rather than $u(.)$.

rowell-17 replied:
corrected `$f(.)$` functions to `$f(\cdot)$`. I was not aware of the correct symbol until now. Thanks!

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+
+This is the reason why SCP can produce an infeasible solution which is referred to as a "soft" failure because usually, a few virtual control terms are nonzero. A feasible solution can be converged through the tuning of the parameters and this leads to the creation of different SCP algorithms such as the SCvx and GuSTO algorithms.
+
+## Main Body
+**SCvx Algorithm**
+The SCvx algorithm follows three things:
+> 1. The terminal and running cost is assumed to be convex by offloading nonconvex terms into the constraints.
+> 2. The constraint (7) is enforced as a hard constraint and $\eta$ is adjusted through the SCvx update rule.
+> 3. Artificial infeasibility is handled using virtual control terms.
+
+**GuSTO Algorithm**
+The GuSTO algorithm follows two things:
+> 1. Artificial unboundedness is handled through the augmentation of the cost function with a soft penalty on the violation of (7) and $\alpha_{u} = 0$ since (1) is convex in the control variables.
+> 2. Artificial infeasibility is handled through the augmentation of the cost function with a soft penalty on nonconvex path constraint violation and as the algorithm progresses, the penalty weight increases.
```
I am looking forward to the Main Body section!

m-kazuki commented on bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md:
```diff
@@ -0,0 +1,127 @@
+$$\delta u(t) = u(t) - \bar{u}(t),$$
+$$\delta p = p - \bar{p},$$
+$$\alpha_{x}\parallel \delta x(t) \parallel_{q} + \alpha_{u}\parallel \delta u(t) \parallel \leq \eta, for t \in [0,1],$$| (7)|
+for some choice of $q \in {1,2,2^{+}, \inf}$ and constants $\alpha_{x}, \alpha_{u}, \alpha_{p} \in {0,1}. Use $q = 2^{+} to denote the ability to impose the trust region as the quadratic two-norm squared. The trust region radius $\eta$ is a fixed scalar that is updated between SCP iterations.
```
This part is not properly enclosed by $.
It should be the following:

for some choice of $q \in {1,2,2^{+}, \inf}$ and constants $\alpha_{x}, \alpha_{u}, \alpha_{p} \in {0,1}$. Use $q = 2^{+}$ to denote the ability to impose the trust region as the quadratic two-norm squared. The trust region radius $\eta$ is a fixed scalar that is updated between SCP iterations.

rowell-17 replied:
Yup, I missed this and I did not double-check. It is fixed now!

m-kazuki reviewed (commented):
From this note, I was able to learn the basics of Sequential Convex Programming. I am also looking forward to the upcoming sections on Successive Convexification and Guaranteed Sequential Trajectory Optimization. I like this note to effectively use diagrams, making it visually easy to understand the concept of SCP. I found the way the equations are presented a little difficult to read. I have provided some specific suggestions below, which I hope you will find helpful.

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

Changed table format of equations to the suggested format

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

corrected $f(.)$ functions to $(\cdot)$

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

incorrect eq.

rowell-17 committed:
Update week-5-e-SCvx-and-GuSTO-SCP.md

(Last update before revise deadline). Added more detail on SCvx. Got stuck with latex math mode formatting issues
